src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,93 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int MAX_ARRAY_CAPACITY = 10;
+    private Object[] values;
+    private int size;
+
+    public ArrayList() {
+        values = new Object[MAX_ARRAY_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        grow(size + 1);
+        values[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index);
+        }
+        grow(size);
+        System.arraycopy(values, index, values, index + 1, size - index);
+        values[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) values[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removedElement = (T) values[index];
+        System.arraycopy(values, index + 1, values, index, size - index - 1);
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (element == values[i] || element != null && element.equals(values[i])) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Element not found: " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void grow(int minCapacity) {
+        if (minCapacity >= values.length) {
```

Comment:
You actually don't need to provide minCapacity here, you can just check 
```suggestion
        if (values.length == size) {
```